README.md
A MangaDex API wrapper for Dart

[![Open in Visual Studio Code](https://open.vscode.dev/badges/open-in-vscode.svg)](https://open.vscode.dev/hernikplays/mangadex-dart-api) [![Dart Test](https://github.com/hernikplays/mangadex-dart-api/actions/workflows/dart.yml/badge.svg)](https://github.com/hernikplays/mangadex-dart-api/actions/workflows/dart.yml)

## Usage

A simple usage example:

```dart
import 'package:mangadex_api/mangadex_api.dart';

main() {
  var client = MDClient();

  client.getChapter('5e8bc984-5f3f-4fb1-b6ee-cf7f3812b112').then((chapter) {
    print(chapter!.title);
  });

  client.getMangaInfo('a96676e5-8ae2-425e-b549-7f15dd34a6d8',
      appendChapters: true, translatedLang: ['cs']).then((m) {
    print(m!.title['en']);
  });
}
```

Currently if you login, only requests __where authentication is required__ will be done with your token, **unless** you set the `useLogin` parameter in a function as true.

## Captcha
If the server returns a captcha, the library will throw a [CaptchaException], which, if handled, has the sitekey inside. After you solve the captcha, you need to pass the result to the [solveCaptcha] function.

## What's implemented
- Log-in
- Getting Chapters
- Getting Manga
- Getting cover
- Manga search
- Getting user
- Captcha
- Getting group
- Logged in user's followed manga chapter feed
- Getting CustomList
- Getting CustomList's chapter feed
- Getting followed manga & group list


## What's NOT yet implemented
- Managing actions (Following, changing reading status etc.)
- Creating accounts
- And other stuff...

**Don't forget to check the [docs](https://pub.dev/documentation/mangadex_api/latest/)**

## Features and bugs

**Please file feature requests and bugs at the [issue tracker][tracker].**

[tracker]: https://github.com/hernikplays/mangadex-dart-api/issues


CHANGELOG.md
## 1.0.3
- Hopefully fixes [#3](https://github.com/hernikplays/mangadex-dart-api/issues/3) for now
- Added `ReadingStatus` class
- Added `roles` to `User`
- Added `followManga` and `unfollowManga`
- Added `setReadingStatus`

For more changes check out the [GitHub diff](https://github.com/hernikplays/mangadex-dart-api/compare/290764345e61dc9906c41dc7f24c5caeb690b808..714c1668d0891a6dabda064b0e788f8b1d6ac84b) and the [documentation][doc]

## 1.0.2
- Fixes [#3](https://github.com/hernikplays/mangadex-dart-api/issues/3)

## 1.0.1
- Add `translatedLanguage` to `getMangaFeed`
- Add reference expansion to `followedManga` and `followedGroups`
- Add `createCustomList`, `deleteCustomList`, `addToCustomList` and `removeFromCustomList`

For more changes check out the [GitHub diff](https://github.com/hernikplays/mangadex-dart-api/compare/d4edb8eb8e40e5b0f0a16cc53031728d988fb3c7..44e5bd1601116fdf38292db6dc9cdf2e1f9ab75a) and the [documentation][doc]

## 1.0.0
- Manga class - cover is now nullable
- Every function should throw exception in case of a 4xx or 5xx error
- Better `getChapter` method, thanks to [#2](https://github.com/hernikplays/mangadex-dart-api/pull/2)
- Functions now validate token using the `validateToken` function
- Added `getUsersList` function
- Added `getListFeed` function
- Added `loggedInUser` function
- Added `followedManga` function
- Added `followedGroups` function


For more changes check out the [GitHub diff](https://github.com/hernikplays/mangadex-dart-api/compare/1.0.0-dev2...1.0.0) and the [documentation][doc]
## 1.0.0-dev.2
- Added User class and getUser function
- Added checking for captcha on request
- added solveCaptcha function
- added Group class and getGroup function
- all requests will **no longer** be automatically sent with your token, if you logged in, you need to set `useLogin` to `true` in the respective function
- added the `mangadex-dart-api/1.0` header to all requests to the API
- **renamed `getMangaInfo` to `getManga`**
- added reference expansion to `getManga` and `search`
- login now **returns a Future**
- **changed in Chapter class: `chapter -> chapterNum` & `volume -> volumeNum`**

For more changes check out the [documentation][doc]

## 1.0.0-dev.1

- Initial version.

[doc]: https://pub.dev/documentation/mangadex_api/latest/


test/normal_test.dart
@Timeout(Duration(seconds: 60))
import 'dart:math';

import 'package:mangadex_api/mangadex_api.dart';
import 'package:test/test.dart';
import 'package:dotenv/dotenv.dart' show load, env;

void main() {
  group('A group of tests', () {
    final client = MDClient();

    setUp(() async {
      load();

      // slow down for rate limit
      var rng = Random();
      await Future.delayed(Duration(seconds: rng.nextInt(2) + 3));
    });

    test('Get Individual Chapter Test', () async {
      var chapter =
          await client.getChapter('5e8bc984-5f3f-4fb1-b6ee-cf7f3812b112');
      expect(chapter!.title, 'Knihovna');
    });

    test('Get Chapter by Manga ID', () async {
      var chapter = await client.getChapter(null,
          mangaId: 'd7037b2a-874a-4360-8a7b-07f2899152fd');
      expect(chapter!.title, "Iruma-kun's Demon School");
    });

    test('Manga Test', () async {
      var manga = await client.getManga('a96676e5-8ae2-425e-b549-7f15dd34a6d8');
      expect(manga!.title['en'], 'Komi-san wa Komyushou Desu.');
    });

    test('Search Test', () async {
      var manga = await client.search(includedTags: [
        '423e2eae-a7a2-4a8b-ac03-a8351462d71d',
        'caaa44eb-cd40-4177-b930-79d3ef2afe87'
      ], status: [
        'ongoing'
      ], authors: [
        '7e552c08-f7cf-4e0e-9723-409d749dd77c'
      ]);
      expect(manga[0].title['jp'], 'Ijiranaide, Nagatoro-san');
    });

    test('Get User Test', () async {
      var user = await client.getUser('b60aca06-048f-4cb9-89c8-87ab2b0dc28f');
      expect(user!.username, 'hernik');
    });

    test('Get Group Test', () async {
      var group = await client.getGroup('790a3272-2a99-4df7-95d1-ee527351a3d0');
      expect(group!.name, 'Weebium');
    });

    test('Search Group Test', () async {
      var group = await client.searchGroups(name: 'Weebium');
      expect(group[0].leader.username, 'hernik');
    });
  });
}


test/logged_in_test.dart
@Timeout(Duration(seconds: 60))
import 'dart:math';

import 'package:mangadex_api/mangadex_api.dart';
import 'package:test/test.dart';
import 'package:dotenv/dotenv.dart' show load, env;

void main() {
  group('Logged-in User test', () {
    final client = MDClient();

    setUp(() async {
      load();

      // slow down for rate limit
      var rng = Random();
      await Future.delayed(Duration(seconds: rng.nextInt(3) + 3));
    });

    test('Followed List Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      var followed = await client.followedManga();
      expect(followed[1].title['en'], 'Tensei Shitara Slime Datta Ken');
    });

    test('Followed Groups Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      var followed = await client.followedGroups();
      expect(followed[0].name, 'Tempest');
      expect(followed[0].leader.roles[2], 'GROUP_LEADER');
    });

    test('Custom List Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      var id = await client.createCustomList('testing list');
      await client.addToCustomList(id, 'c2390196-0ad7-4b90-9019-c23c000eec78');
      await Future.delayed(Duration(seconds: 2));
      await client.removeFromCustomList(
          id, 'c2390196-0ad7-4b90-9019-c23c000eec78');
      await client.deleteCustomList(id);
    });

    test('Update Manga Status Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      await client.setReadingStatus(
          'e78a489b-6632-4d61-b00b-5206f5b8b22b', ReadingStatus.PAUSED);
    });

    test('Follow & Unfollow manga test', () async {
      await client.login('4lomega', env['MDPASS']!);
      await client.followManga('e78a489b-6632-4d61-b00b-5206f5b8b22b');
      await Future.delayed(Duration(seconds: 3));
      await client.unfollowManga('e78a489b-6632-4d61-b00b-5206f5b8b22b');
    });
  });
}


test/mangadex_api_test.dart
@Timeout(Duration(seconds: 60))
import 'dart:math';

import 'package:mangadex_api/mangadex_api.dart';
import 'package:test/test.dart';
import 'package:dotenv/dotenv.dart' show load, env;

void main() {
  group('A group of tests', () {
    final client = MDClient();

    setUp(() async {
      load();

      // slow down for rate limit
      var rng = Random();
      await Future.delayed(Duration(seconds: rng.nextInt(2) + 3));
    });

    test('Get Individual Chapter Test', () async {
      var chapter =
          await client.getChapter('5e8bc984-5f3f-4fb1-b6ee-cf7f3812b112');
      expect(chapter!.title, 'Knihovna');
    });

    test('Get Chapter by Manga ID', () async {
      var chapter = await client.getChapter(null,
          mangaId: 'd7037b2a-874a-4360-8a7b-07f2899152fd');
      expect(chapter!.title, "Iruma-kun's Demon School");
    });

    test('Manga Test', () async {
      var manga = await client.getManga('a96676e5-8ae2-425e-b549-7f15dd34a6d8');
      expect(manga!.title['en'], 'Komi-san wa Komyushou Desu.');
    });

    test('Search Test', () async {
      var manga = await client.search(includedTags: [
        '423e2eae-a7a2-4a8b-ac03-a8351462d71d',
        'caaa44eb-cd40-4177-b930-79d3ef2afe87'
      ], status: [
        'ongoing'
      ], authors: [
        '7e552c08-f7cf-4e0e-9723-409d749dd77c'
      ]);
      expect(manga[0].title['jp'], 'Ijiranaide, Nagatoro-san');
    });

    test('Get User Test', () async {
      var user = await client.getUser('b60aca06-048f-4cb9-89c8-87ab2b0dc28f');
      expect(user!.username, 'hernik');
    });

    test('Get Group Test', () async {
      var group = await client.getGroup('790a3272-2a99-4df7-95d1-ee527351a3d0');
      expect(group!.name, 'Weebium');
    });

    test('Search Group Test', () async {
      var group = await client.searchGroups(name: 'Weebium');
      expect(group[0].leader.username, 'hernik');
    });

    test('Logged In User - Followed List Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      var followed = await client.followedManga();
      expect(followed[0].title['jp'], 'Tensei Shitara Slime Datta Ken');
    });

    test('Logged In User - Followed Groups Test', () async {
      await client.login('4lomega', env['MDPASS']!);
      var followed = await client.followedGroups();
      expect(followed[0].name, 'Tempest');
    });
  });
}


lib/src/classes.dart
/// Holds data about a single Chapter
class Chapter {
  /// Chapter ID
  final String id;

  /// Title of the chapter
  final String title;

  /// Volume number, is [Null] if not specified in API
  final String? volumeNum;

  /// Chapter number, is [Null] if not specified in API
  final String? chapterNum;

  /// Language of the chapter
  final String translatedLanguage;

  /// URLs for individual pages
  final List<String> chapterURLs;

  /// URLs for pages with data-saver enabled
  final List<String> dataSaverChapterURLs;

  /// The uploader of the chapter
  final String? uploader;

  /// IDs of groups credited in the chapter
  final List<String>? groupIds;

  /// ID of the parent manga
  final String? mangaId;

  /// Date when the chapter was created
  final String createdAt;

  /// Date of last chapter update
  final String updatedAt;

  Chapter(
      {required this.dataSaverChapterURLs,
      this.uploader,
      this.groupIds,
      this.mangaId,
      required this.createdAt,
      required this.updatedAt,
      required this.id,
      required this.title,
      required this.translatedLanguage,
      required this.chapterURLs,
      this.volumeNum,
      this.chapterNum});
}

/// Holds data about a Manga
class Manga {
  /// Manga ID
  final String id;

  /// Original title of manga
  final Map<String, dynamic> title;

  /// Alternative manga titles (with other languages)
  final List<dynamic> altTitles;

  /// Full manga description
  final Map<String, dynamic> description;

  /// Manga locked status
  final bool? isLocked;

  /// Links to author etc.
  final Map<String, dynamic>? links;

  /// Original language of the manga
  final String originalLang;

  /// Last volume
  final String? lastVolume;

  /// Last chapter
  final String? lastChapter;

  /// Target demographic
  final String? demographic;

  /// Publishing status
  final String? status;

  /// Year of release
  final String? releaseYear;

  /// Content rating
  final String? contentRating;

  /// Manga tags
  final List<dynamic> tags;

  /// Date when the chapter was created
  final String createdAt;

  /// Date of last chapter update
  final String updatedAt;

  /// available chapters from the /aggregate endpoint
  final Map<String, dynamic>? chapters;

  /// the cover art, which is currently in use on the website
  final String? cover;

  /// Manga author
  final Author author;

  /// Manga artist
  final Author artist;

  Manga(
      {required this.id,
      required this.title,
      required this.altTitles,
      required this.description,
      required this.isLocked,
      required this.originalLang,
      required this.tags,
      required this.createdAt,
      required this.updatedAt,
      this.demographic,
      this.lastChapter,
      this.lastVolume,
      this.contentRating,
      this.links,
      this.releaseYear,
      this.status,
      this.chapters,
      this.cover,
      required this.artist,
      required this.author});
}

/// Holds data about a MangaDex user
class User {
  /// User's username
  final String username;

  /// User's ID
  final String id;

  /// User's roles
  final List<String> roles;

  User({required this.username, required this.id, required this.roles});
}

/// Exception thrown when the API server returns a 403 error requiring completion of a captcha
///
/// For more info on MangaDex API's captchas check [this part of their API](https://api.mangadex.org/docs.html#section/Captchas)
///
/// For sending the captcha result use the solveCaptcha function
class CaptchaException implements Exception {
  /// Additional message
  final String message;

  /// Recaptcha site key used for rendering the [widget](https://developers.google.com/recaptcha/docs/display#auto_render)
  final String sitekey;
  CaptchaException(this.sitekey, {this.message = ''});
}

/// Holds data about a Scanlation group
class Group {
  /// Name of the group
  final String name;

  /// Group ID
  final String id;

  /// Group description
  final String? description;

  /// The group's leader user
  final User leader;

  /// Group's set website
  final String? website;

  /// Group's set IRC server
  final String? ircServer;

  /// Group's set IRC channel
  final String? ircChannel;

  /// Group's set discord
  final String? discord;

  /// Group's set e-mail
  final String? contactEmail;

  /// When the group was created
  final String createdAt;

  /// When the group was last updated
  final String updatedAt;

  /// Whether the group is locked
  final bool isLocked;

  /// All group members (including leaders)
  final List<User>? members;

  Group(
      {required this.name,
      required this.id,
      this.description,
      required this.leader,
      this.website,
      this.ircServer,
      this.ircChannel,
      this.discord,
      this.contactEmail,
      required this.createdAt,
      required this.updatedAt,
      required this.isLocked,
      this.members});
}

/// A manga author OR artist
class Author {
  /// Author's/Artist's name
  final String name;

  /// Author's/Artist's biographies
  final List<dynamic> biography;

  /// Author's/Artist's id
  final String id;

  /// Author's/Artist's image
  final String? imageUrl;

  Author(
      {required this.name,
      required this.biography,
      required this.id,
      this.imageUrl});
}

/// Anything that does not have additional attributes than these
class GenericObject {
  /// Object's ID
  String id;

  /// Supplied title
  String? title;

  /// Object's type
  String? type;

  GenericObject({required this.id, this.title, this.type});
}

/// All MangaDex reading statuses
class ReadingStatus {
  static final String READING = 'reading';
  static final String PAUSED = 'on_hold';
  static final String PLANNING = 'plan_to_read';
  static final String DROPPED = 'dropped';
  static final String REREADING = 're_reading';
  static final String COMPLETED = 'completed';
}


example/mangadex_api_example.dart
import 'package:mangadex_api/mangadex_api.dart';

void main() {
  // Example usage of some functions
  var client = MDClient();
  client.login('user', 'pass').then((v) {
    client
        .getChapter(null, mangaId: 'd7037b2a-874a-4360-8a7b-07f2899152fd')
        .then((chapter) {
      print(chapter!.title);
    });

    client.getManga('a96676e5-8ae2-425e-b549-7f15dd34a6d8',
        appendChapters: true, translatedLang: ['cs']).then((m) {
      print(m!.title['en']);
    });

    client.getCovers('a96676e5-8ae2-425e-b549-7f15dd34a6d8').then((covers) {
      print(covers![covers.length - 1]);
    });

    client.search(
        mangaTitle: 'Nagatoro',
        authors: ['7e552c08-f7cf-4e0e-9723-409d749dd77c']).then((res) {
      print(res[0].title);
    });

    client.getGroup('790a3272-2a99-4df7-95d1-ee527351a3d0').then((group) {
      print(group!.name);
    });

    client.searchGroups(name: 'Weebium').then((group) {
      print(group[0].leader.username);
    });

    client.getMangaFeed().then((value) {
      print(value[0].chapterNum);
    });
  });
}
